### PR TITLE
[3338] Added basic top level course stats

### PR DESCRIPTION
**Pull request by defong**
### Context
Vital statistics
 
### Changes proposed in this pull request
Add basic course stats

### Guidance to review
Subject stats will have dupes as a course can be associated to one or more subjects
http://localhost:3001/reporting

```json
{
  "total": {
    "all": 17575,
    "non_findable": 4626,
    "all_findable": 12949
  },
  "findable_total": {
    "open": 9886,
    "closed": 3063
  },
  "provider_type": {
    "open": {
      "scitt": 1641,
      "lead_school": 7257,
      "university": 987,
      "unknown": 0,
      "invalid_value": 0
    },
    "closed": {
      "scitt": 332,
      "lead_school": 2637,
      "university": 94,
      "unknown": 0,
      "invalid_value": 0
    }
  },
  "program_type": {
    "open": {
      "higher_education_programme": 1035,
      "school_direct_training_programme": 6213,
      "school_direct_salaried_training_programme": 1068,
      "scitt_programme": 1493,
      "pg_teaching_apprenticeship": 77
    },
    "closed": {
      "higher_education_programme": 101,
      "school_direct_training_programme": 1784,
      "school_direct_salaried_training_programme": 807,
      "scitt_programme": 305,
      "pg_teaching_apprenticeship": 66
    }
  },
  "study_mode": {
    "open": {
      "full_time": 9304,
      "part_time": 178,
      "full_time_or_part_time": 404
    },
    "closed": {
      "full_time": 2960,
      "part_time": 32,
      "full_time_or_part_time": 71
    }
  },
  "qualification": {
    "open": {
      "qts": 1890,
      "pgce_with_qts": 7842,
      "pgde_with_qts": 118,
      "pgce": 30,
      "pgde": 6
    },
    "closed": {
      "qts": 653,
      "pgce_with_qts": 2391,
      "pgde_with_qts": 18,
      "pgce": 0,
      "pgde": 1
    }
  },
  "is_send": {
    "open": {
      "yes": 95,
      "no": 9791
    },
    "closed": {
      "yes": 50,
      "no": 3013
    }
  },
  "subject": {
    "open": {
      "Primary": 1281,
      "Primary with English": 14,
      "Primary with geography and history": 11,
      "Primary with mathematics": 133,
      "Primary with modern languages": 20,
      "Primary with physical education": 35,
      "Primary with science": 13,
      "Art and design": 325,
      "Science": 15,
      "Biology": 627,
      "Business studies": 164,
      "Chemistry": 693,
      "Citizenship": 21,
      "Classics": 12,
      "Communication and media studies": 25,
      "Computing": 475,
      "Dance": 51,
      "Design and technology": 440,
      "Drama": 266,
      "Economics": 30,
      "English": 725,
      "Geography": 609,
      "Health and social care": 26,
      "History": 508,
      "Mathematics": 843,
      "Music": 314,
      "Philosophy": 0,
      "Physical education": 215,
      "Physics": 750,
      "Psychology": 62,
      "Religious education": 406,
      "Social sciences": 58,
      "French": 331,
      "English as a second or other language": 1,
      "German": 161,
      "Italian": 4,
      "Japanese": 5,
      "Mandarin": 13,
      "Russian": 3,
      "Spanish": 279,
      "Further education": 36,
      "Modern Languages": 781,
      "Modern languages (other)": 305
    },
    "closed": {
      "Primary": 421,
      "Primary with English": 1,
      "Primary with geography and history": 1,
      "Primary with mathematics": 28,
      "Primary with modern languages": 1,
      "Primary with physical education": 5,
      "Primary with science": 1,
      "Art and design": 140,
      "Science": 4,
      "Biology": 178,
      "Business studies": 84,
      "Chemistry": 149,
      "Citizenship": 0,
      "Classics": 12,
      "Communication and media studies": 11,
      "Computing": 113,
      "Dance": 28,
      "Design and technology": 124,
      "Drama": 67,
      "Economics": 5,
      "English": 226,
      "Geography": 138,
      "Health and social care": 13,
      "History": 185,
      "Mathematics": 152,
      "Music": 95,
      "Philosophy": 0,
      "Physical education": 323,
      "Physics": 142,
      "Psychology": 33,
      "Religious education": 89,
      "Social sciences": 47,
      "French": 114,
      "English as a second or other language": 1,
      "German": 70,
      "Italian": 2,
      "Japanese": 0,
      "Mandarin": 4,
      "Russian": 1,
      "Spanish": 100,
      "Further education": 1,
      "Modern Languages": 254,
      "Modern languages (other)": 75
    }
  }
}
```

shamelessly borrowed #1345 

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/controllers/reporting_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      open: column == :provider_type ? open.transform_keys { |key| Provider.provider_types.key(key || "") } : open,
+      closed: column == :provider_type ? closed.transform_keys { |key| Provider.provider_types.key(key || "") } : closed,
+    }
+  end
+
+  def build_courses
+    @courses = @recruitment_cycle.courses
+    @findable_courses = @courses.findable.distinct
+    @open_courses = @findable_courses.with_vacancies
+    @closed_courses = @findable_courses.where.not id: @open_courses
```
Comment (dankmitchell) [inline app/controllers/reporting_controller.rb:48]:
```suggestion
    @closed_courses = @findable_courses.where.not(id: @open_courses)
```

File: spec/requests/reporting_spec.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+require "rails_helper"
+
+describe "GET /reporting" do
+  it "returns status success" do
+    get "/reporting"
+    expect(response.status).to eq(200)
```
Comment (dankmitchell) [inline spec/requests/reporting_spec.rb:6]: I think we need to test the outcome of the JSON here...

File: app/controllers/reporting_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        closed: @closed_courses.count,
+      },
+      provider_type: { **group_by_count(:provider_type) },
+      program_type: { **group_by_count(:program_type) },
+
+      study_mode: { **group_by_count(:study_mode) },
+      qualification: { **group_by_count(:qualification) },
+      is_send: { **group_by_count(:is_send) },
+    }
+  end
```
Comment (misaka) [inline app/controllers/reporting_controller.rb:31]: I think this should go into a service or reporting class. Then you can break down the tests to one per stat, and keep them out of the request / feature specs.

**Comment by defong** [comment]
[[S] Top-level stats endpoint - courses](https://trello.com/c/g0z6yocY/3338-s-top-level-stats-endpoint-courses)

File: spec/services/course_reporting_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        closed: {
+          qts: 0, pgce_with_qts: 0, pgde_with_qts: 0, pgce: 0, pgde: 0
+        },
+      },
+      is_send: {
+        open: { yes: 1, no: 2 },
+        closed:  { yes: 0, no: 0 },
+      },
+      subject: {
+        open: Subject.active.each_with_index.map { |sub, i| x = {}; x[sub.subject_name] = (i + 1) * 3; x }.reduce({}, :merge),
```
Comment (asmega) [inline spec/services/course_reporting_service_spec.rb:80]: i think you cheated and copied the implementation in to the test. why did you choose to do that here and have the actual outputs on the other parts of the test?

Comment (defong) [reply spec/services/course_reporting_service_spec.rb:80]: The whole test is a cheat...

The input is a fake the in between is also a fake. 

The bit I'm really testing for in this whole thing is really


```ruby

{
 "I want this key to be transformed" : "either a number or defaulted as 0" 
} 

```

**Pull request merged**